src/frame/endpoints/GroupApiEndpoints.ts

```diff
@@ -272,14 +286,45 @@ export default {
     /**
      * Eventually this will allow us to get a filtered list of groups based on the provided list of group IDs. But for now it's
      * just getting the entire list of groups available to the current user.
-     * @param {string[]} groupIDs List of group IDs to retrieve
+     * @param {string[]} groupIds List of group IDs to retrieve
      */
-    callGroupKeyListApi(groupIDs: string[]): Future<SDKError, GroupListResponseType> {
-        if (!groupIDs.length) {
+    callGroupKeyListApi(groupIds: string[]): Future<SDKError, GroupListResponseType> {
+        if (!groupIds.length) {
             return Future.of({result: []});
         }
-        const {url, options, errorCode} = groupList(groupIDs);
-        return makeAuthorizedApiRequest<GroupListResponseType>(url, errorCode, options);
+        const {url, options, errorCode} = groupList(groupIds);
+        return makeAuthorizedApiRequest<GroupListResponseType>(url, errorCode, options).map((groupKeyListResponse) => {
+            // cache off the public keys for everything we just requests
+            groupKeyListResponse.result.forEach((fullGroupResponse) => {
+                groupPublicKeyCache[fullGroupResponse.id] = {id: fullGroupResponse.id, groupMasterPublicKey: fullGroupResponse.groupMasterPublicKey};
+            });
+            return groupKeyListResponse;
+        });
+    },
+
+    getGroupPublicKeyList(groupIds: string[]): Future<SDKError, GroupPublicKeyObject[]> {
+        if (!groupIds.length) {
+            return Future.of([]);
+        }
+
+        // try to get keys from the cache
+        const cacheHits: GroupPublicKeyObject[] = [];
+        groupIds.forEach((groupId) => {
+            const cacheValue = groupPublicKeyCache[groupId];
+            if (cacheValue) {
+                cacheHits.push(cacheValue);
+            }
+        });
+
+        // if all the group ids were cached, then we should just return those
+        // otherwise call through to the API (which will update the cache)
+        if (cacheHits.length === groupIds.length) {
+            return Future.of(cacheHits);
+        } else {
+            return this.callGroupKeyListApi(groupIds).map((groupListResponse) =>
+                groupListResponse.result.map(({id, groupMasterPublicKey}) => ({id, groupMasterPublicKey}))
+            );
+        }
```

Comment:
So is this only adding benefit if all the keys were previously cached? I was expecting the `else` to only make calls for groupIds that weren't in `cacheHits` and then return the combined list of the calls and the hits. Unless I'm misunderstanding something? Or is this our means of updating the cache?

Reply:
It only adds benefits if all keys were already cached. I did write exactly the code you're talking about initially, but it was significantly more complicated and I was told my the rest of the team that it didn't provide benefit: if we're making the call, we might as well call for all of them.